bitswap/client/internal/peermanager/peerwantmanager.go

```diff
@@ -152,7 +221,65 @@ func (pwm *peerWantManager) broadcastWantHaves(wantHaves []cid.Cid) {
 		if len(peerUnsent) > 0 {
 			pws.peerQueue.AddBroadcastWantHaves(peerUnsent)
 		}
+
+		if bcastLimitPeers > 0 {
+			bcastLimitPeers--
+			if bcastLimitPeers == 0 {
+				break
+			}
+		}
+	}
+}
+
+func (pwm *peerWantManager) skipBroadcast(peerID peer.ID, peerQueue PeerQueue) bool {
+	// Broadcast to peer from which block(s) have been previously received.
+	pwm.bcastMutex.Lock()
+	_, ok := pwm.bcastTargets[peerID]
+	pwm.bcastMutex.Unlock()
+	if ok {
+		return false
+	}
+
+	// Broadcast to peers on local network.
+	if pwm.isLocalPeer(peerID) {
+		// Add local peer to broadcast targets to avoid next isLocalPeer check.
+		pwm.bcastMutex.Lock()
+		pwm.bcastTargets[peerID] = struct{}{}
+		pwm.bcastMutex.Unlock()
+		return false
+	}
+
+	// Broadcast to peers that have a pending message to piggyback on.
+	if pwm.bcastConfig.SendWithPending && peerQueue.HasMessage() {
+		return false
+	}
+	return true
+}
+
+func (pwm *peerWantManager) markBroadcastTarget(peerID peer.ID) {
+	if pwm.bcastTargets == nil {
+		return
+	}
+	pwm.bcastMutex.Lock()
+	pwm.bcastTargets[peerID] = struct{}{}
+	pwm.bcastMutex.Unlock()
+}
+
+func (pwm *peerWantManager) isLocalPeer(peerID peer.ID) bool {
+	if pwm.bcastConfig.SendLocalPeers == nil {
+		return false
+	}
+	if _, ok := pwm.remotePeers[peerID]; ok {
+		return false
+	}
+	addrs := pwm.bcastConfig.SendLocalPeers.Addrs(peerID)
+	for _, addr := range addrs {
+		if manet.IsPrivateAddr(addr) {
```

Comment:
this is not enough, we also want to include loopback addresses to avoid the need for https://github.com/ipfs/helia/pull/818 

```suggestion
		if manet.IsPrivateAddr(addr) || manet.IsIPLoopback(addr) {
```

Reply:
Thank you! I had assumed that loopback was a private address, but apparently not

Reply:
If the intent is all non public addresses then maybe ! isPublic is better?

cc @sukunrt for any thoughts, and also it might be to helpful in manet to make the sugaring for working with the various address types here easier / more obvious.